src/wifi/model/periodic-dmg-wifi-scheduler.cc

```diff
@@ -0,0 +1,490 @@
+        {
+          continue;
+        }
+
+      if (newEndAlloc < startSlot)
+        {
+          // the following snippet ensures that the search is carried on until 
+          // we find the empty gap created by the allocation time reduction
+          // and we add it to the list of available slots 
+          if (keepSearching)
+            {
+              if (difference == (startSlot - newEndAlloc))
+                {
+                  // two adjacent available slots are merged 
+                  it->first = newEndAlloc;
+                  keepSearching = false;
+                  break;
+                }
+              else if (difference < (startSlot - newEndAlloc))
+                {
+                  // this condition is verified when there is one or more allocations 
+                  // between the gap and the following available slot
+                  m_availableSlots.insert (it, std::make_pair (newEndAlloc, newEndAlloc + difference));
+                  keepSearching = false;
+                  break;
+                }
+              NS_ASSERT_MSG (difference <= (startSlot - newEndAlloc), "Something broke in runtime, check the update of the available slots.");
+            }
+        }
+    }
+
+  m_remainingDtiTime += difference;
+
+  for (const auto & slot : m_availableSlots)
+    {
+      NS_LOG_DEBUG ("Available slot from " << slot.first << " to " << slot.second);
+    }
+
+}
+
+StatusCode
+PeriodicDmgWifiScheduler::ModifyExistingAllocation (uint8_t sourceAid, const DmgTspecElement &dmgTspec, const DmgAllocationInfo &info)
+{
+  NS_LOG_FUNCTION (this << +sourceAid);
+
+  StatusCode status;
+  uint32_t newDuration;
+  if (info.GetAllocationFormat () == ISOCHRONOUS)
+    {
+      newDuration = GetAllocationDuration (dmgTspec.GetMinimumAllocation (), dmgTspec.GetMaximumAllocation ());
+    }
+  else if (info.GetAllocationFormat () == ASYNCHRONOUS)
+    {
+      // for asynchronous allocations, the Maximum Allocation field is reserved (IEEE 802.11ad 8.4.2.136) 
+      newDuration = dmgTspec.GetMinimumAllocation ();
+    }
+  else
+    {
+      NS_FATAL_ERROR ("Allocation Format not supported");
+    }
+
+  AllocationFieldListI allocation;
+  // Retrieve the allocation for which a modification has been requested
+  for (allocation = m_addtsAllocationList.begin (); allocation != m_addtsAllocationList.end ();)
+    {
+      if ((allocation->GetAllocationID () == info.GetAllocationID ())
+          && (allocation->GetSourceAid () == sourceAid) 
+          && (allocation->GetDestinationAid () == info.GetDestinationAid ()))
+        {
+          break;
+        }
+      else
+        {
+          ++allocation;
+        }
+    }
+
+  NS_ABORT_MSG_IF (allocation == m_addtsAllocationList.end (), "Required allocation does not exist.");
+
+  uint32_t currentDuration = allocation->GetAllocationBlockDuration ();
+  NS_LOG_DEBUG ("current duration=" << currentDuration << ", new duration=" << newDuration);
+  uint32_t timeDifference;
```

Comment:
In the case of Isochronous traffic, we should check also the minimum allocation duration before aborting the modification request. Maybe the minimum allocation value is less then the current duration.